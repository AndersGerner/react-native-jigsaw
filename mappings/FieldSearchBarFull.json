[{"name":"Search Bar","tag":"FieldSearchBarFull","description":"A search bar with accompanying search icon and clear button.","category":"field","preview_image_url":"https://res.cloudinary.com/altos/image/upload/draftbit/library/jigsaw-1.0/reps/Field_SearchBar_Full.png","supports_list_render":false,"props":{"icon":{"label":"Icon","description":"Left icon to display","type":"icon","value":null,"editable":true,"required":false},"placeholder":{"label":"Placeholder","description":"Input placeholder text","type":"string","value":"Type something...","editable":true,"required":false},"onSubmit":{"label":"Submit action","description":"Action to execute on submission","editable":true,"type":"action","value":null},"fieldName":{"label":"Field name","description":"The name of the field within the screen that will store this component's value","type":"fieldName","value":"searchBarValue","valuePropName":"value","handlerPropName":"onChange","editable":true,"required":false}},"layout":{}}]