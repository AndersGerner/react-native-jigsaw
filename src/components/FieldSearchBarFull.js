/* @flow */

import * as React from "react";
import { View, Text, TextInput, StyleSheet } from "react-native";
import { withTheme } from "../core/theming";
import {
  COMPONENT_TYPES,
  FORM_TYPES,
  FIELD_NAME
} from "../core/component-types";
import Icon from "./Icon";
import Config from "./Config";
import type { Theme } from "../types";

type Props = {
  icon: string,
  placeholder?: string,
  onSubmit: (text: string) => void,
  theme: Theme,
  style?: any
};

class FieldSearchBarFull extends React.Component<Props> {
  state = {
    focused: false
  };

  static defaultProps = {
    icon: "search"
  };

  clearText = () => {
    this.setState({ value: "" });
  };

  onBlur = () => {
    this.setState({ focused: false });
  };

  onChange = value => {
    this.props.onChange && this.props.onChange(value);
  };

  onFocus = () => {
    this.setState({ focused: true });
  };

  onSubmit = () => {
    this.props.onSubmit && this.props.onSubmit();
  };

  render() {
    const {
      icon,
      placeholder,
      onSubmit,
      style,
      theme: { colors, spacing, typography },
      onChange,
      value
    } = this.props;

    const { focused } = this.state;

    const { lineHeight, ...typeStyle } = typography.body2;

    return (
      <View style={[{ padding: spacing.large }, styles.container, style]}>
        <Icon
          name={icon}
          size={Config.fieldSearchBarFullIconSize}
          color={focused ? colors.primary : colors.light}
        />
        <TextInput
          clearButtonMode="while-editing"
          placeholder={placeholder}
          value={value}
          onBlur={this.onBlur}
          onFocus={this.onFocus}
          onChange={onChange}
          onSubmitEditing={this.onSubmit}
          placeholderTextColor={colors.light}
          style={[
            typeStyle,
            {
              color: colors.medium,
              marginLeft: spacing.medium,
              flex: 1,
              justifyContent: "center"
            }
          ]}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center"
  }
});

export default withTheme(FieldSearchBarFull);

export const SEED_DATA = [
  {
    name: "Search Bar",
    tag: "FieldSearchBarFull",
    description: "A search bar with accompanying search icon and clear button.",
    category: COMPONENT_TYPES.field,
    preview_image_url: "{CLOUDINARY_URL}/Field_SearchBar_Full.png",
    supports_list_render: false,
    props: {
      icon: {
        label: "Icon",
        description: "Left icon to display",
        type: FORM_TYPES.icon,
        value: null,
        editable: true,
        required: false
      },
      placeholder: {
        label: "Placeholder",
        description: "Input placeholder text",
        type: FORM_TYPES.string,
        value: "Type something...",
        editable: true,
        required: false
      },
      onSubmit: {
        label: "Submit action",
        description: "Action to execute on submission",
        editable: true,
        type: FORM_TYPES.action,
        value: null
      },
      fieldName: FIELD_NAME
    },
    layout: {
      width: 375,
      height: 56
    }
  }
];
