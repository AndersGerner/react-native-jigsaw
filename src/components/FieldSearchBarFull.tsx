import * as React from "react";
import {
  View,
  TextInput,
  StyleSheet,
  StyleProp,
  ViewStyle,
  NativeSyntheticEvent,
  TextInputChangeEventData,
  TextInputSubmitEditingEventData,
} from "react-native";
import { withTheme } from "../core/theming";
import {
  GROUPS,
  COMPONENT_TYPES,
  FORM_TYPES,
  PROP_TYPES,
  FIELD_NAME,
} from "../core/component-types";
import Icon from "./Icon";
import Config from "./Config";
import theme from "../styles/DefaultTheme";

interface Props {
  icon?: string;
  placeholder?: string;
  style?: StyleProp<ViewStyle>;
  theme: typeof theme;
  onChange?: (e: NativeSyntheticEvent<TextInputChangeEventData>) => void;
  onSubmit?: (e: NativeSyntheticEvent<TextInputSubmitEditingEventData>) => void;
  value: string;
}

const FieldSearchBarFull: React.FC<Props> = ({
  icon = "search",
  placeholder = "",
  style,
  theme: { colors, spacing, typography },
  onChange: changeOverride,
  onSubmit: submitOverride,
  value,
}) => {
  const [focused, setIsFocused] = React.useState(false);

  const onBlur = () => {
    setIsFocused(false);
  };

  const onChange = (e: NativeSyntheticEvent<TextInputChangeEventData>) => {
    changeOverride && changeOverride(e);
  };

  const onFocus = () => {
    setIsFocused(true);
  };

  const onSubmit = (
    e: NativeSyntheticEvent<TextInputSubmitEditingEventData>
  ) => {
    submitOverride && submitOverride(e);
  };

  const { ...typeStyle } = typography.body2;

  return (
    <View style={[{ padding: spacing.large }, styles.container, style]}>
      <Icon
        name={icon}
        size={Config.fieldSearchBarFullIconSize}
        color={focused ? colors.primary : colors.light}
      />
      <TextInput
        clearButtonMode="while-editing"
        placeholder={placeholder}
        value={value}
        onBlur={onBlur}
        onFocus={onFocus}
        onChange={onChange}
        onSubmitEditing={onSubmit}
        placeholderTextColor={colors.light}
        style={[
          typeStyle,
          {
            color: colors.medium,
            marginLeft: spacing.medium,
            flex: 1,
            justifyContent: "center",
          },
        ]}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center",
  },
});

export default withTheme(FieldSearchBarFull);

export const SEED_DATA = [
  {
    name: "Search Bar",
    tag: "FieldSearchBarFull",
    description: "A search bar with accompanying search icon and clear button.",
    category: COMPONENT_TYPES.field,
    preview_image_url: "{CLOUDINARY_URL}/Field_SearchBar_Full.png",
    supports_list_render: false,
    props: {
      icon: {
        group: GROUPS.basic,
        label: "Icon",
        description: "Left icon to display",
        formType: FORM_TYPES.icon,
        propType: PROP_TYPES.ASSET,
        defaultValue: null,
        editable: true,
        required: false,
      },
      placeholder: {
        group: GROUPS.basic,
        label: "Placeholder",
        description: "Input placeholder text",
        formType: FORM_TYPES.string,
        propType: PROP_TYPES.STRING,
        defaultValue: "Search for...",
        editable: true,
        required: false,
      },
      onSubmit: {
        group: GROUPS.basic,
        label: "Submit action",
        description: "Action to execute on submission",
        editable: true,
        required: false,
        formType: FORM_TYPES.action,
        propType: PROP_TYPES.STRING,
        defaultValue: null,
      },
      fieldName: {
        ...FIELD_NAME,
        defaultValue: "searchBarValue",
      },
    },
    layout: {},
  },
];
