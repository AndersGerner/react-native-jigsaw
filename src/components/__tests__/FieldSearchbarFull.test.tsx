import * as React from "react";
import renderer from "react-test-renderer";
import FieldSearchBarFull from "../FieldSearchBarFull";

it("renders field search bar full", () => {
  const tree = renderer
    .create(<FieldSearchBarFull onChange={() => {}} value="Test Value" />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
