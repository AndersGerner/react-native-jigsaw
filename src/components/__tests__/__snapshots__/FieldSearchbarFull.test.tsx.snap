// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders field search bar full 1`] = `
<View
  style={
    Array [
      Object {
        "padding": 16,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "marginLeft": 12,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      clearButtonMode="while-editing"
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder=""
      placeholderTextColor="rgba(165, 173, 183, 1)"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "color": "rgba(70, 78, 88, 1)",
          },
          Object {
            "backgroundColor": "transparent",
            "fontFamily": "System",
            "fontSize": 14,
            "fontWeight": "400",
            "letterSpacing": 0,
          },
        ]
      }
      underlineColorAndroid="transparent"
      value="Test Value"
    />
  </View>
</View>
`;
